main.cpp

```diff
@@ -1,48 +1,574 @@
+    FloatType& subtract(float f);
+    FloatType& multiply(float f);
+    FloatType& divide(float f);
+
+    FloatType& add(const IntType& it);
+    FloatType& add(const FloatType& ft);
+    FloatType& add(const DoubleType& dt);
+
+    FloatType& subtract(const IntType& it);
+    FloatType& subtract(const FloatType& ft);
+    FloatType& subtract(const DoubleType& dt);
+
+    FloatType& multiply(const IntType& it);
+    FloatType& multiply(const FloatType& ft);
+    FloatType& multiply(const DoubleType& dt);
+
+    FloatType& divide(const IntType& it);
+    FloatType& divide(const FloatType& ft);
+    FloatType& divide(const DoubleType& dt);
+};
+
+struct DoubleType
+{
+    double* value = nullptr;
+    DoubleType(double doublePrimitive);
+    ~DoubleType();
+
+    DoubleType& add(double d);
+    DoubleType& subtract(double d);
+    DoubleType& multiply(double d);
+    DoubleType& divide(double d);
+
+    DoubleType& add(const IntType& it);
+    DoubleType& add(const FloatType& ft);
+    DoubleType& add(const DoubleType& dt);
+
+    DoubleType& subtract(const IntType& it);
+    DoubleType& subtract(const FloatType& ft);
+    DoubleType& subtract(const DoubleType& dt);
+
+    DoubleType& multiply(const IntType& it);
+    DoubleType& multiply(const FloatType& ft);
+    DoubleType& multiply(const DoubleType& dt);
+
+    DoubleType& divide(const IntType& it);
+    DoubleType& divide(const FloatType& ft);
+    DoubleType& divide(const DoubleType& dt);    
+};
+
+struct IntType
+{
+    int* value = nullptr;
+    IntType(int IntPrimitive);
+    ~IntType();
+
+    IntType& add(int i);
+    IntType& subtract(int i);
+    IntType& multiply(int i);
+    IntType& divide(int i);
+
+    IntType& add(const IntType& it);
+    IntType& add(const FloatType& ft);
+    IntType& add(const DoubleType& dt);
+
+    IntType& subtract(const IntType& it);
+    IntType& subtract(const FloatType& ft);
+    IntType& subtract(const DoubleType& dt);
+
+    IntType& multiply(const IntType& it);
+    IntType& multiply(const FloatType& ft);
+    IntType& multiply(const DoubleType& dt);
+
+    IntType& divide(const IntType& it);
+    IntType& divide(const FloatType& ft);
+    IntType& divide(const DoubleType& dt);
+};
+
+/* FloatType member function definitions */
+FloatType::FloatType(float floatPrimitive) 
+{
+    value = new float(floatPrimitive);
+}
+
+FloatType::~FloatType() 
+{
+    delete value;
+}
+
+FloatType& FloatType::add(float f)
+{
+    *value += f;
+    return *this;
+}
+
+FloatType& FloatType::subtract(float f)
+{
+    *value -= f;
+    return *this;
+}
+
+FloatType& FloatType::multiply(float f)
+{
+    *value *= f;
+    return *this;
+}
+
+FloatType& FloatType::divide(float f)
+{
+    if (f == 0.0f)
+    {
+        std::cout << "warning: floating point division by zero!" << std::endl;
+        *value /= std::numeric_limits<float>::denorm_min();
+        return *this;
+    }
+    else
+    {
+        *value /= f;
+        return *this;
+    }
+}
```

Comment:
D.R.Y.  both blocks return `*this`
You changed the functionality from part1's divide function.
continue to divide by your function argument.   Warn if `f == 0.f` only.
- [ ] revise.
```
if( f == 0.f )
    issue the warning...

do the division
return *this;
```